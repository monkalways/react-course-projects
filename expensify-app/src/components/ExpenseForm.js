import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import { Button, Form, Input, TextArea } from 'semantic-ui-react';

export default class ExpenseForm extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            id: props.expense ? props.expense.id : '',
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }

    onDescriptionChange = (e, data) => {
        const description = data.value;
        this.setState(() => ({ description }));
    };
    onNoteChange = (e, data) => {
        const note = data.value;
        this.setState(() => ({ note }));
    };
    onAmountChange = (e, data) => {
        const amount = data.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) { // validate number
            this.setState(() => ({ amount }));
        }
    };
    onDateChange = (createdAt) => {
        if(createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    onCalendarFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };
    onSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount' }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                id: this.state.id,
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: +this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    }
    render() {
        return (
            <Form onSubmit={this.onSubmit}>
                {this.state.error || <p>{this.state.error}</p>}
                <Form.Field>
                    <label>Description</label>
                    <Input 
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                </Form.Field>
                <Form.Field>
                    <label>Amount</label>
                    <Input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                </Form.Field>
                <Form.Field>
                    <label>Created At</label>
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onCalendarFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day) => false}
                        showDefaultInputIcon
                    />
                </Form.Field>
                <Form.Field>
                    <label>Note</label>
                    <TextArea
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </TextArea>
                </Form.Field>
                <Button type="submit" primary>Save Expense</Button>
            </Form>
        );
    }
}
