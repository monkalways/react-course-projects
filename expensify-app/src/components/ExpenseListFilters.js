import React from 'react';
import { connect } from 'react-redux';

import { setTextFilter, sortByAmount, sortByDate } from '../actions/filters';

const ExpenseListFilters = ({ filters, dispatch }) => {
    return (
        <div>
            <input 
                type="text" 
                value={filters.text} 
                onChange={(e) => {
                    const text = e.target.value;
                    dispatch(setTextFilter(text));
                }} 
            />
            <select 
                value={filters.sortBy} 
                onChange={(e) => {
                    const filterBy = e.target.value;
                    if (filterBy === 'date') {
                        dispatch(sortByDate());
                    } else if (filterBy === 'amount') {
                        dispatch(sortByAmount());
                    }
                }}
            >
                <option value="date">Date</option>
                <option value="amount">Amount</option>
            </select>
        </div>
    );
};

const mapStateToProps = (state) => ({
    filters: state.filters
});

export default connect(mapStateToProps)(ExpenseListFilters);
