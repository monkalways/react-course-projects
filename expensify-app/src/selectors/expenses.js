// Get visible expenses
const selectExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter(expense => {
        const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
        const textMatch = !text || expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        switch (sortBy) {
            case 'date':
                return a.createdAt < b.createdAt ? 1 : -1;
            case 'amount':
                return a.amount < b.amount ? 1 : -1;
            default:
                return -1;
        }
    });
};

export default selectExpenses;
