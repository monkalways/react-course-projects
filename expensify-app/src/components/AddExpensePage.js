import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';
import PageHeader from './PageHeader';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        const { startAddExpense, history } = this.props;
        startAddExpense(expense);
        history.push('/');
    };

    render() {
        return (
            <div>
                <PageHeader>
                    <h1>Add Expense</h1>
                </PageHeader>
                <ExpenseForm onSubmit={this.onSubmit} />
            </div> 
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);