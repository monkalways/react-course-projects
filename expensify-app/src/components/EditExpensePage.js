import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

const EditExpensePage = ({ expense, dispatch, history }) => {
    return (
        <div>
            <ExpenseForm
                expense={expense}
                onSubmit={(expense) => {
                    console.log(expense);
                    dispatch(editExpense(expense.id, expense));
                    history.push('/');
                }} 
            />
            <button onClick={() => {
                dispatch(removeExpense({ id: expense.id }));
                history.push('/');
            }}>Remove</button>
        </div>
    );
};

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find(expense => expense.id === props.match.params.id)
    };
};

export default connect(mapStateToProps)(EditExpensePage);