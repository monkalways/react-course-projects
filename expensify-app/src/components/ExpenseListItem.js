import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
import { Header, Table } from 'semantic-ui-react';

import { removeExpense } from '../actions/expenses';

export const ExpenseListItem = ({ id, description, amount, createdAt }) => (
    <Table.Row>
        <Table.Cell>
            <Link to={`/edit/${id}`}>
                <Header as="h3">
                    {description}
                    <Header.Subheader>
                        {moment(createdAt).format('MMMM Do, YYYY')}
                    </Header.Subheader>
                </Header>
            </Link>
        </Table.Cell>
        <Table.Cell>
            <Header as="h3">{numeral(amount/100).format('$0,0.00')}</Header>
        </Table.Cell>
    </Table.Row>
);

export default ExpenseListItem;