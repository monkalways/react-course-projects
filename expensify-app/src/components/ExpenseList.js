import React from 'react';
import { connect } from 'react-redux';
import { Table } from 'semantic-ui-react';

import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = ({ expenses }) => (
    <div>
        {
            expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                <Table stackable>
                    <Table.Header>
                    <Table.Row>
                        <Table.HeaderCell>Expense</Table.HeaderCell>
                        <Table.HeaderCell>Amount</Table.HeaderCell>
                    </Table.Row>
                    </Table.Header>
                    <Table.Body>
                        {expenses.map(expense => (
                            <ExpenseListItem key={expense.id} {...expense} />
                        ))}
                    </Table.Body>
                </Table>
                
            )
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);