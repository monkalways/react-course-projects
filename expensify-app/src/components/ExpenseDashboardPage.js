import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';

import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import getVisibleExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';

export const ExpenseDashboardPage = ({ expenses, expensesTotal }) => {
    const count = expenses ? expenses.length : 0;
    const expenseWording = count > 1 ? 'expenses' : 'expense';
    const formattedTotal = numeral(expensesTotal/100).format('$0,0.00');
    return (
        <div>
            <p>{`Viewing ${count} ${expenseWording} totalling ${formattedTotal}`}</p>
            <ExpenseListFilters />
            <ExpenseList />
        </div>
    )
};

const mapStateToProps = (state) => ({
    expenses: getVisibleExpenses(state.expenses, state.filters),
    expensesTotal: getExpensesTotal(getVisibleExpenses(state.expenses, state.filters))
});

export default connect(mapStateToProps)(ExpenseDashboardPage);