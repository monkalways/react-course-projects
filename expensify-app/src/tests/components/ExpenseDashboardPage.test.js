import 'react-dates/initialize' // to make react-dates work
import React from 'react';
import { shallow } from 'enzyme';

import { ExpenseDashboardPage}  from '../../components/ExpenseDashboardPage';
import expenses from '../fixtures/expenses';

test('should render ExpenseDashboardPage with single expense', () => {
    const expensesTotal = 195;
    const wrapper = shallow(<ExpenseDashboardPage expenses={[expenses[0]]} expensesTotal={expensesTotal} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseDashboardPage with multiple expenses', () => {
    const expensesTotal = 114195;
    const wrapper = shallow(<ExpenseDashboardPage expenses={expenses} expensesTotal={expensesTotal} />);
    expect(wrapper).toMatchSnapshot();
});