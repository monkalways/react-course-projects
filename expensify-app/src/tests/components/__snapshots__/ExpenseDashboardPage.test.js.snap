// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseDashboardPage with multiple expenses 1`] = `
<div>
  <p>
    Viewing 3 expenses totalling $1,141.95
  </p>
  <Connect(ExpenseListFilters) />
  <Connect(ExpenseList) />
</div>
`;

exports[`should render ExpenseDashboardPage with single expense 1`] = `
<div>
  <p>
    Viewing 1 expense totalling $1.95
  </p>
  <Connect(ExpenseListFilters) />
  <Connect(ExpenseList) />
</div>
`;
